### PR TITLE
Added SQS interactions for S3 source

### LINKED ISSUE #1425: Add basic SQS interactions
1. Use the SQS ReceiveMessage API to receive messages from SQS.
2. For each Message from SQS, it will:
- Parse the Message as an S3Event.
 - Download the S3 Object which the S3Event indicates was created.
 - Decompress the object if configured to do so.
 - Parse the decompressed file using the configured codec into a list of Log Event objects.
 - Write the Log objects into the Data Prepper buffer.

3. Perform a DeleteMessageBatch with all of the messages which were successfully processed.
Repeat

### PR DESCRIPTION
**Pull request by asifsmohammed**
### Description

- Added sqs interaction to `getMessage` from SQS
- And convert each Message to `S3EventNotificationRecord`
- Added exponential backoff for reading messages from SQS

Used AWS SDK v1 for Deserializing SQS message into `S3EventNotificationRecord` since it's not yet ported to v2. For more info see: https://github.com/aws/aws-sdk-java-v2/issues/1197
 
### Issues Resolved
Resolves #1425 
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/s3-source/build.gradle [hunk 1]
```diff
@@ -13,10 +13,13 @@ repositories {
 
 dependencies {
     implementation project(':data-prepper-api')
-    implementation 'com.fasterxml.jackson.core:jackson-databind'
+    implementation project(':data-prepper-plugins:blocking-buffer')
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/build.gradle:16]: Why did you add `blocking-buffer` here? Data Prepper API provides the `Buffer` interface.

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/SqsWorker.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                        .queueUrl(sqsOptions.getSqsUrl())
+                        .maxNumberOfMessages(sqsOptions.getMaximumMessages())
+                        .visibilityTimeout((int) sqsOptions.getVisibilityTimeout().getSeconds())
+                        .waitTimeSeconds((int) sqsOptions.getWaitTime().getSeconds())
+                        .build();
+
+                messages.addAll(sqsClient.receiveMessage(receiveMessageRequest).messages());
+            } catch (SqsException e) {
+                LOG.error(e.awsErrorDetails().errorMessage());
+                System.exit(1);
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/SqsWorker.java:57]: This will exit all of Data Prepper. That is a far too severe consequence of an error communicating with SQS. I think that we want to log the exception, then wait some time, then try again. It could be appropriate to perform an exponential backoff.

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/SqsWorker.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                try {
+                    Thread.sleep(s3SourceConfig.getSqsOptions().getPollDelay().toMillis());
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+            }
+        } while (true);
+    }
+
+    private S3EventNotification.S3EventNotificationRecord getS3EventMessages(Message message) {
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/SqsWorker.java:90]: I recommend renaming this so that it has a different verb than "get." Perhaps "convert" or "transform?" Get makes it seem that you are using a getter.

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/configuration/SqsOptions.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import jakarta.validation.constraints.Max;
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotBlank;
+
+import java.time.Duration;
+
+public class SqsOptions {
+    private final int DEFAULT_MAXIMUM_MESSAGES = 10;
+    private final int DEFAULT_VISIBILITY_TIMEOUT_SECONDS = 30;
+    private final int DEFAULT_WAIT_TIME_SECONDS = 0;
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/configuration/SqsOptions.java:18]: I think this default should be 20 seconds (the current maximum).

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/configuration/SqsOptions.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import jakarta.validation.constraints.Max;
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotBlank;
+
+import java.time.Duration;
+
+public class SqsOptions {
+    private final int DEFAULT_MAXIMUM_MESSAGES = 10;
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/configuration/SqsOptions.java:16]: These should be `private static final` (They are missing the `static`).

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/configuration/SqsOptions.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    private final int DEFAULT_VISIBILITY_TIMEOUT_SECONDS = 30;
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/configuration/SqsOptions.java:17]: Your choice here: You could make these defaults actual `Duration` objects. Since they will be static, you won't need to re-create the `Duration` in each field. I also think it is cleaner since the code only deals with `Duration`.

e.g.

```
private static final Duration DEFAULT_VISIBILITY_TIMEOUT = Duration.ofSeconds(20);
```

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Source.java [hunk 7]
```diff
@@ -13,44 +13,57 @@
...
     @Override
     public void stop() {
 
     }
 
-    private final AwsCredentialsProvider awsCredentialsProvider;
-    private final S3Client s3Client;
-
-    private AwsCredentialsProvider createCredentialsProvider() {
+    public AwsCredentialsProvider createCredentialsProvider() {
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Source.java:66]: Why is this public and not private?

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Source.java [hunk 8]
```diff
@@ -13,44 +13,57 @@
...
     }
 
     @Override
     public void start(Buffer<Record<Event>> buffer) {
+        if (buffer == null) {
+            throw new IllegalStateException("Buffer provided is null");
+        }
 
+        for(int i = 0; i < s3SourceConfig.getSqsOptions().getThreadCount(); i++) {
+            Thread sqsWorkerThread = new Thread(new SqsWorker(sqsClient, s3Client, s3SourceConfig));
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Source.java:56]: I think we will want to have a `List<Thread>` as a field. You can keep all the threads there. You will need to stop them when the `stop()` method for S3Source is called.

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/SqsWorker.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                ReceiveMessageRequest receiveMessageRequest = ReceiveMessageRequest.builder()
+                        .queueUrl(sqsOptions.getSqsUrl())
+                        .maxNumberOfMessages(sqsOptions.getMaximumMessages())
+                        .visibilityTimeout((int) sqsOptions.getVisibilityTimeout().getSeconds())
+                        .waitTimeSeconds((int) sqsOptions.getWaitTime().getSeconds())
+                        .build();
+
+                messages.addAll(sqsClient.receiveMessage(receiveMessageRequest).messages());
+            } catch (SqsException e) {
+                LOG.error(e.awsErrorDetails().errorMessage());
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/SqsWorker.java:56]: In general you should avoid logging just an error message from some external source. It is better to have:

```
LOG.error("Error reading from SQS: {}", e.awsErrorDetails().errorMessage());
```

Relatedly - is `awsErrorDetails()` guaranteed to be non-null?

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/SqsWorker.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+            // TODO: Write to buffer
+
+
+            //TODO: delete messages which are successfully processed
+
+            if (messages.size() < sqsOptions.getMaximumMessages() && s3SourceConfig.getSqsOptions().getPollDelay().toMillis() > 0) {
+                try {
+                    Thread.sleep(s3SourceConfig.getSqsOptions().getPollDelay().toMillis());
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/SqsWorker.java:84]: We should log this, not call `e.printStackTrace()`.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/s3-source/build.gradle (hunk 1 above)
Comment (asifsmohammed) [reply data-prepper-plugins/s3-source/build.gradle:16]: I added this for testing S3Source using buffer.

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/configuration/SqsOptions.java (hunk 6 above)
Comment (asifsmohammed) [reply data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/configuration/SqsOptions.java:17]: I like this suggestion.

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/SqsWorker.java (hunk 9 above)
Comment (asifsmohammed) [reply data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/SqsWorker.java:56]: I'm not sure if `awsErrorDetails()` will be non-null. I'll check this one

File: data-prepper-plugins/s3-source/build.gradle (hunk 1 above)
Comment (sbayer55) [reply data-prepper-plugins/s3-source/build.gradle:16]: For testing could this import be added as `testImplementation project(':data-prepper-plugins:blocking-buffer')`?

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/BackoffUtils.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+package com.amazon.dataprepper.plugins.source;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Random;
+
+public class BackoffUtils {
```
Comment (sbayer55) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/BackoffUtils.java:13]: In place of utils could `BackoffUtils` class have a more descriptive name? Consider something like `RetryState` or `BackoffManager`. Without looking at the class file I expected BackoffUtils to contain some helper functions for parsing or transforming data.

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/BackoffUtils.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    public boolean shouldRetry() {
+        return numberOfRetries > 0;
+    }
+
+    public void errorOccurred() {
+        numberOfRetries -= 1;
+
+        if (shouldRetry()) {
+            waitUntilNextTry();
+            timeToWait += random.nextInt(1000);
```
Comment (sbayer55) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/BackoffUtils.java:35]: It there a benefit to a random wait opposed to a fixed time or exponential growth time?

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3ClientAuthentication.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    private AwsAuthenticationOptions awsAuthenticationOptions;
+
+    public S3ClientAuthentication(final S3SourceConfig s3SourceConfig) {
+        this.s3SourceConfig = s3SourceConfig;
+        this.awsAuthenticationOptions = s3SourceConfig.getAWSAuthentication();
+    }
+
+    public S3Client createS3Client(final StsClient stsClient) {
+
+        return software.amazon.awssdk.services.s3.S3Client.builder()
```
Comment (sbayer55) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3ClientAuthentication.java:24]: Can the fully qualified name be shortened from `software.amazon.awssdk.services.s3.S3Client` to `S3Client`? It looks like you already have the import.

**Review by sbayer55 — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/s3-source/build.gradle (hunk 1 above)
Comment (dlvenable) [reply data-prepper-plugins/s3-source/build.gradle:16]: @sbayer55 's suggestion is the right approach for a testing dependency.

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/BackoffUtils.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (dlvenable) [reply data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/BackoffUtils.java:13]: We do need a different mechanism in place for the entire thread. As the SQS processing runs in its own thread, it must not end until the the source is shutdown.

I think we should use both:
* AWS SDK retry to help with transient errors
* Catch exceptions in the thread, suppress, and wait. This goes beyond AWS SDK errors. Say, some message in the SQS queue is not actually JSON. We don't want the thread to fail for a single bad message.

That being said, perhaps this will not need any sophisticated back-off. Perhaps a straightforward `Thread.sleep()` is enough.

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Source.java [hunk 14]
```diff
@@ -12,52 +12,43 @@
...
-    private AwsCredentialsProvider createCredentialsProvider() {
-        return Objects.requireNonNull(s3SourceConfig.getAWSAuthentication().authenticateAwsConfiguration(StsClient.create()));
-    }
-
-    private S3Client createS3Client(final AwsCredentialsProvider awsCredentialsProvider) {
-        return S3Client.builder()
-                .region(Region.of(s3SourceConfig.getAWSAuthentication().getAwsRegion()))
-                .credentialsProvider(awsCredentialsProvider)
-                .build();
+        Thread.currentThread().interrupt();
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Source.java:52]: Data Prepper core is calling the `stop()` method. Thus, this current code is going to interrupt the Data Prepper core thread.

Also, rather than use `interrupt()`, I'd like to have a `running`/`shutdown` boolean flag in the worker thread. You then can use `while(running) { ... }` instead of `while(true) { ... }`.

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/BackoffUtils.java (hunk 12 above)
Comment (asifsmohammed) [reply data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/BackoffUtils.java:35]: Yeah, it's to break the synchronization across the clients thereby avoiding collisions

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/SqsWorker.java [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+            }
+
+            // read each message as S3 event message
+            Map<Message, S3EventNotification.S3EventNotificationRecord> s3EventNotificationRecords = messages.stream()
+                    .collect(Collectors.toMap(message -> message, this::convertS3EventMessages));
+
+            // build s3ObjectPointer from S3EventNotificationRecord if event name starts with ObjectCreated
+            List<S3ObjectReference> addedObjects = new ArrayList<>();
+            for (Map.Entry<Message, S3EventNotification.S3EventNotificationRecord> entry: s3EventNotificationRecords.entrySet()) {
+                if (entry.getValue().getEventName().startsWith("ObjectCreated")) {
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/SqsWorker.java:59]: I'd like to create an `S3EventFilter` structure. There are couple motivations I have in this:

1. We are also going to add an optional user-configuration to support filtering by bucketName or accountId.
2. If we add SNS as an option in future iterations, this filter should be the same code.


Here is a possible design:

```
public interface S3EventFilter {
  Optional<S3EventNotification.S3EventNotificationRecord> filter(S3EventNotification.S3EventNotificationRecord notification)
}

class ObjectCreatedFilter implements S3EventFilter {
  Optional<S3EventNotification.S3EventNotificationRecord> filter(S3EventNotification.S3EventNotificationRecord entry) {
    if (entry.getValue().getEventName().startsWith("ObjectCreated"))
      return Optional.of(entry);
    else
      return Optional.empty();
  }
}
```

We can build more complicated filters for the bucketName and accountId in another PR.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3ObjectReference.java [hunk 16]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.plugins.source;
+
+import java.util.Objects;
+
+public class S3ObjectReference {
```
Comment (dlvenable) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3ObjectReference.java:10]: You might get a merge conflict on this file since I also added this into `main`. Please use the one I provided as it has `toString()` and is unit tested.

**Review by dlvenable — DISMISSED** [review]

**Review by dlvenable — APPROVED** [review]

File: data-prepper-plugins/s3-source/src/test/java/com/amazon/dataprepper/plugins/source/SqsWorkerTest.java [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+import software.amazon.awssdk.services.sqs.SqsClient;
+import software.amazon.awssdk.services.sqs.model.ReceiveMessageRequest;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+class SqsWorkerTest {
```
Comment (cmanning09) [inline data-prepper-plugins/s3-source/src/test/java/com/amazon/dataprepper/plugins/source/SqsWorkerTest.java:22]: The `run` method is not tested. We are missing on out some core test coverage should add test cases.

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/SqsWorker.java [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+    public SqsWorker(final SqsClient sqsClient, final S3Client s3Client, final S3SourceConfig s3SourceConfig) {
+        this.s3SourceConfig = s3SourceConfig;
+        this.sqsClient = sqsClient;
+        this.s3Client = s3Client;
+        objectMapper = new ObjectMapper();
+        sqsOptions = s3SourceConfig.getSqsOptions();
+    }
+
+    @Override
+    public void run() {
```
[the remark this answers is not in the document]
Comment (cmanning09) [reply data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/SqsWorker.java:45]: This will also make it easier to unit test and can remove variable reassignment/mutation in the code below.

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/SqsWorker.java [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class SqsWorker implements Runnable {
+    private static final Logger LOG = LoggerFactory.getLogger(SqsWorker.class);
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    private S3SourceConfig s3SourceConfig;
```
Comment (cmanning09) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/SqsWorker.java:30]: These private variables should be final to prevent re-assignment

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Service.java [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+    private S3Client s3Client;
+
+    public S3Service(final S3SourceConfig s3SourceConfig) {
+        this.s3SourceConfig = s3SourceConfig;
+        this.s3Client = createS3Client(StsClient.create());
+    }
+
+    S3ObjectReference addS3Object(final S3ObjectReference s3ObjectReference) {
+        // TODO: should return message id and receipt handle if successfully converted to event
+        return null;
```
Comment (cmanning09) [inline data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Service.java:29]: I have a few concerns over this method and subsequent workflow. I have some ideas for improvements but I want to understand a few things first.

What are we adding an S3Object to?  I find this method name confusing and was really confused by the usage in the `run()` method.

Why are the parameters and return object the same object (S3ObjectReference)? Are we extending S3ObjectReference and expecting this class to mutate the variable with a messageId and Receipt? (I want to see if we can avoid mutation and nullable variables)

How is the receipt used by the calling function? Do we need it?

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Source.java [hunk 21]
```diff
@@ -12,52 +12,38 @@
...
-
-        awsCredentialsProvider = createCredentialsProvider();
-        s3Client = createS3Client(awsCredentialsProvider);
-
-        throw new NotImplementedException();
     }
 
     @Override
     public void start(Buffer<Record<Event>> buffer) {
+        if (buffer == null) {
```
[the remark this answers is not in the document]
Comment (dlvenable) [reply data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Source.java:35]: The `S3ObjectWorker` handles this:

https://github.com/opensearch-project/data-prepper/blob/be35437fe4174647babed4169e6890e2b1fed5a2/data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3ObjectWorker.java#L30

Also, @asifsmohammed and I are coordinating on the work here. He created this PR before the `S3ObjectWorker` was present in `main`, so some portions like this are currently unused in his branch and work.

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Service.java (hunk 20 above)
Comment (dlvenable) [reply data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Service.java:29]: > Why are the parameters and return object the same object (S3ObjectReference)? Are we extending S3ObjectReference and expecting this class to mutate the variable with a messageId and Receipt? (I want to see if we can avoid mutation and nullable variables)

@asifsmohammed and I discussed a couple approaches here:
1. We can make this templated and add something like `T getNotificationMetadata()` which in this case returns SQS metadata. It could be extended for SNS metadata if that were ever added.
2. We could extend this class with a sub-class.

In either case, the SQS metadata would include the `ReceiptHandle` and `MessageId`.

I'm fine with either approach. But, the inheritance approach would have the benefit of completely abstracting this metadata from the S3 code.

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Source.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (asifsmohammed) [reply data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Source.java:35]: I'll update the logic with buffer here

File: data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Service.java (hunk 20 above)
Comment (asifsmohammed) [reply data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Service.java:29]: My initial thought was to add `ReceiptHandle` and `MessageId` to `S3ObjectReference` which will be used to get S3 Object.
If S3 object is successfully processed we return the same `S3ObjectReference` to delete message from queue. 

I'll create a subclass which contains `ReceiptHandle` and `MessageId`.

**Review by cmanning09 — APPROVED** [review]
Thanks for making those changes. Let's address the remaining in the next PR.

**Review by dlvenable — APPROVED** [review]
